### ISSUE TITLE
RFE: warn about fragile (or all?) ${#array[@]} uses

### ISSUE THREAD

**Issue by scop** [koalaman/shellcheck#1929]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

Using `${#array[@]}` to derive the last valid index in array is fragile; the array might have "holes" in it in form of unset values. Would be nice if shellcheck would warn about this.

One case is looping over an array and using the above construct in the initializer or the condition.

Another that has a code smell is when `${#array[@]}-1` is used for anything really.

Perhaps even warn about using `${#array[@]}` in the first place.

Many of these can be fixed by just looping over the array values like
```bash
for x in "${array[@]}"; do ...
```
, or if indices are needed, looping over the assigned ones like
```bash
for i in ${!array[@]}; do ...`.
```

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
foo=(a b poof c last)
unset "foo[2]"

# does not loop to the end -- does not echo "last"
for (( i=0; i < ${#foo[@]}; i++ )); do
    echo "${foo[i]}"
done

echo ===
# neither does this
i=0
while [[ $i -lt ${#foo[@]} ]]; do
    echo "${foo[i]}"
    (( i++ ))
done

echo ===
# does not loop from the end -- does not echo "last"
for (( i=${#foo[@]}-1; i >= 0; i-- )); do
    echo "${foo[i]}"
done

echo ===
# assumption that index of last element in array is ${#foo[@]}-1 is not true
echo "last element is NOT ${foo[${#foo[@]}-1]}!"
```

#### Here's what shellcheck currently says:

Nothing.

#### Here's what I wanted or expected to see:

```
for (( i=0; i < ${#foo[@]}; i++ )); do
                ^--------^ SCnnn: some warning here

while [[ $i -lt ${#foo[@]} ]]; do
                ^--------^ SCnnn: some warning here

for (( i=${#foo[@]}-1; i >= 0; i-- )); do
         ^----------^ SCnnn: some warning here

echo "last element is NOT ${foo[${#foo[@]}-1]}!"
                                ^----------^ SCnnn: some warning here
```

**Comment by brother** [comment]
The second "expected warning" (the while statement) is a bit far fetched imho. The others are on target for a warning.

If you can rely on the script context as a whole it might pick up that a var is an array and has been manipulated between two points in time (like in the example provided). At that point all warnings bells are welcome. In that case I would imagine no warnings on the other hand if there are no unsets between defining and using the element counter though.
